### PR TITLE
#268 #269 Improve HTTPX Support

### LINKED ISSUE #268: HTTPX OAuth1 implementation does not set Content-Length header
**Describe the bug**
The OAuth1 implementation for HTTPX does not set the content-length header, which can cause problems when using body signatures and talking to servers who block body POSTs where the length header is not specified.

**Expected behavior**

Content-Length header is set.

**Environment:**

 - OS: linux
 - Python Version: 3.8
 - Authlib Version: 0.14.3

Incoming PR fixes this.

### LINKED ISSUE #269: Support for HTTPX Synchronous API
**Is your feature request related to a problem? Please describe.**

HTTPX has synchronous APIs as well as async, authlib should support both.

**Additional context**

Incoming PR supplies this.

### PR DESCRIPTION
**Pull request by berzoidberg**
Fixes #268
Fixes #269  

**What kind of change does this PR introduce?** (check at least one)

- [x] Bugfix
- [x] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Other, please describe:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

- [x] You consent that the copyright of your pull request source code belongs to Authlib's author.

### CODE REVIEW

**Comment by lepture** [comment]
@berzoidberg I'm not ready to support sync mode of httpx. currently httpx is not stable, it has many breaking changes. For example, I have locked httpx version to `httpx==0.14.1`, because `0.14.2` will break our code. I'd like to fix test cases for `httpx==0.14.2` at first.

**Comment by berzoidberg** [comment]
I can get that done. 0.14.3 is the latest, should I aim to get tests working with that version instead?

**Comment by berzoidberg** [comment]
@lepture Successfully updated to work w/HTTPX 0.14.3 -- the newest available.

**Comment by lepture** [comment]
Actually, we had HTTPX sync API before, in version 0.13. But later after we released 0.13, HTTPX removed sync API for a while.

**Comment by berzoidberg** [comment]
Aye, in HTTPX 0.13.0 they released the new sync client, sync and async both use the same codebase with some minor interface differences now. I'll gladly volunteer to continue maintaining HTTPX support as they make changes if we can get this merged up.

File: authlib/integrations/httpx_client/assertion_client.py [hunk 1]
```diff
@@ -1,4 +1,5 @@
-from httpx import AsyncClient
+from httpx import AsyncClient, Client
+from httpx._config import UnsetType
```
Comment (lepture) [inline authlib/integrations/httpx_client/assertion_client.py:2]: Is there a way to not using private module in httpx?

File: authlib/integrations/httpx_client/assertion_client.py [hunk 2]
```diff
@@ -51,3 +52,47 @@ async def _refresh_token(self, data):
             )
         self.token = token
         return self.token
+
+class AssertionClient(_AssertionClient, Client):
```
Comment (lepture) [inline authlib/integrations/httpx_client/assertion_client.py:56]: 2 lines between class.

File: authlib/integrations/httpx_client/assertion_client.py [hunk 3]
```diff
@@ -51,3 +52,47 @@ async def _refresh_token(self, data):
...
+    DEFAULT_GRANT_TYPE = JWT_BEARER_GRANT_TYPE
+
+    def __init__(self, token_endpoint, issuer, subject, audience=None, grant_type=None,
+                 claims=None, token_placement='header', scope=None, **kwargs):
+
+        client_kwargs = extract_client_kwargs(kwargs)
+        Client.__init__(self, **client_kwargs)
+
+        _AssertionClient.__init__(
+            self, session=None,
```
Comment (lepture) [inline authlib/integrations/httpx_client/assertion_client.py:71]: if `session=self`, there is no need to re-write `_refresh_token`.

File: authlib/integrations/httpx_client/oauth2_client.py [hunk 4]
```diff
@@ -150,3 +151,101 @@ def _http_post(self, url, body=None, auth=None, headers=None, **kwargs):
...
+            self.refresh_token(url, refresh_token=refresh_token)
+        elif self.metadata.get('grant_type') == 'client_credentials':
+            access_token = self.token['access_token']
+            token = self.fetch_token(url, grant_type='client_credentials')
+            if self.update_token:
+                self.update_token(token, access_token=access_token)
+        else:
+            raise InvalidTokenError()
+
+    def _fetch_token(self, url, body='', headers=None, auth=None,
```
Comment (lepture) [inline authlib/integrations/httpx_client/oauth2_client.py:212]: You can check the implementation of `requests`, there is no need to implemented it if you have set `session=self`.

**Comment by berzoidberg** [comment]
@lepture thanks for the review, i removed the unnecessary bits from all 3 client types (assertion, oauth1, oauth2)

**Comment by berzoidberg** [comment]
@lepture After reviewing the way you have the code organized, I thought it best to move the most reusable components of the async clients into the oauth1 and oauth2 directories, rather than have it live solely inside the httpx clients. Let me know if you agree this is the correct move -- it does require the use of a version-check in order to support older version of python, but I think that's "ok"

**Pull request merged**